h2o-core/src/main/java/water/fvec/Frame.java

```diff
@@ -1619,13 +1622,47 @@ public CSVStream(Chunk[] chks, String[] names, int nChunks, CSVStreamParams parm
       StringBuilder sb = new StringBuilder();
       if (names != null) {
         sb.append('"').append(names[0]).append('"');
-        for(int i = 1; i < names.length; i++)
+        for (int i = 1; i < names.length; i++)
           sb.append(_parms._separator).append('"').append(names[i]).append('"');
         sb.append('\n');
       }
       _line = StringUtils.bytesOf(sb);
       _chkRow = -1; // first process the header line
       _curChks = chks;
+      _escapedCategoricalVecDomains = escapeCategoricalVecDomains(_curChks);
+    }
+
+    /**
+     * Escapes categorical levels of vectors and puts them in a map of escaped categorical levels.
+     * Only the domains with at least one level with an escaped quote are saved. If a domain does not need
+     * any escaping, it is considered better practice to reach to the `vec.domain()` method itself and not duplicate entries
+     * in memory here.
+     *
+     * @param chunks
+     * @return A 2D array of String[][]. Elements can be null of give domain does not need escaping.
+     */
+    private String[][] escapeCategoricalVecDomains(final Chunk[] chunks) {
+      final String[][] localEscapedCategoricalVecDomains = new String[chunks.length][];
+
+      for (int i = 0; i < chunks.length; i++) {
+        final Vec vec = chunks[i].vec();
+        if (!vec.isCategorical()) continue;
+
+        final String[] originalDomain = vec.domain();
+        final String[] escapedDomain = new String[originalDomain.length];
+
+        boolean escapingRequired = false;
```

Comment:
This adds complexity and doesn't IMHO address any issue. You can simply keep the original domain and avoid switching between escaped/original in method getEscapedCategoricalLevel.

Reply:
Got rid of the null checking in `getEscapedCategoricalLevel` - this is definitely unnecessary complexity.

But I can't see how to drop the `escapingRequired` - at the end, we need to either drop the escaped domain if it's the same and use reference to the original `vec.domain()` or use the escaped domain. And we need something to base the decision on, and that is the `escapingRequired`. 

Doing `final String[] escapedDomain = vec.domain()` and then modifying it would mean the matcher would modify the original domain. Or we'd need to duplicate it, which goes against our aim to save memory.

Reply:
Looks good to me now, thank you